_posts/2020-01-09-rxjs-wstep.md

```diff
@@ -0,0 +1,180 @@
+### map
+Map jest z pewnością znanym Ci operatorem, chociażby z API JS Arrays.map, którym RxJS się sugerował. Map transformuje dane zwracając nam nowy wynik dla każdej emitowanej danej. Przykładowo:
+```typescript
+observable$.pipe(map(numerek => numerek*2))
+    .subscribe(value => console.log('zmapowana wartosc: ', value))
+// zmapowana wartosc: 30000
+```
+
+### first
+Interesującym operatorem jest operator _first()_, który pobiera pierwszy element ze strumienia, następnie jako efekt uboczny - wypisuje obserwatora z listy obserwujących dany strumień! Jest to bardzo wygodne, jeżeli z danego gorącego strumienia potrzebujemy tylko jednej wartości - nie musimy przejmować się w tym momencie wywołaniem funkcji unsubscribe().
+Przykład:
+```typescript
+observable$.pipe(first())
+    .subscribe(value => console.log('pierwsza otrzymana wartosc: ', value))
+// pierwsza otrzymana wartosc: 15000
+```
+### withLatestFrom
+Następnym operatorem wartym uwagi jest _withLatestFrom_, który umożliwia nam 'skrzyżowanie strumieni', czyli dołączenie do jednego strumienia ostatnio emitowanej wartości przez drugi strumień. Przykładowo:
+```typescript
+const otherStream$ = new BehaviourSubject<boolean>(true);
+observable$.pipe(withLatestFrom(otherStream$))
+    .subscribe(([value, otherStreamValue]) => console.log('otrzymana wartosc z pierwszego strumienia: {}, z drugiego: {}', value, otherStreamValue))
+// otrzymana wartosc z pierwszego strumienia: 15000, z drugiego: true
+```
+### takeUntil
+Operator takeUntil jest wygodnym operatorem, jeżeli chcemy w efekciarski i prosty sposób wypisać się z listy obserwujących dany strumień.
+Rzućmy okiem na kod:
+```typescript
+@Component({
+  selector: 'app-some-component',
+  template: `
+        some template
+  `,
+  styleUrls: ['some-styles.scss']
+})
+export class SomeComponent implements OnDestroy, OnInit {
+    private destroySubject$ = new Subject<void>();
+    private someValues$ = new BehaviourSubject<string>('initial value');
+    
+    ngOnInit(): void {
+       this.someValues$.asObservable().pipe(takeUnitl(this.destroySubject$))
+            .subscribe(value => ...)
+    }
+
+    ngOnDestroy(): void {
+       this.destroySubject.next();
+    }   
+
+}
+``` 
+Powyższy fragment przedstawia nam SomeComponent, który nasłuchuje na wartości ze strumienia _someValues$_ aż do momentu, gdy zostanie wyemitowana wartość z drugiego strumienia: _destroySubject$_. W taki oto sposób nie musimy pamiętać o ręcznym odsubskrybowaniu się z pierwszego strumienia, gdyż zostanie on automatycznie zamknięty w momencie zniszczenia komponentu przez Angulara. :)
+### switchMap
+Kolejnym przydatnym operatorem może być _switchMap()_, którego zastosowanie pozwala nam zachować czystość i powstrzyma nas przed tworzeniem tzw. _callback hell_, czyli zagnieżdżania wywołań _subscribe()_ tworzących łańcuch wywołań trudny w czytaniu i utrzymywaniu. Dlatego też zamiast pisać tak:
```

Comment:
To jest niepoprawne językowo (można coś zrobić _tak po prostu_ lub zrobić coś _w taki sposób_, ale nie zrobić coś tak)
Dlatego też zamiast pisać tak:
==> 
Dlatego też zamiast pisać:

Reply:
poprawione